File: backtrack_cpp/src/kinematics.cpp
#include "backtrack/kinematics.hpp"

#include <cmath>
#include <stdexcept>

#include "backtrack/geometry.hpp"

namespace backtrack {

namespace {
constexpr double kTwoPi = 2.0 * M_PI;
constexpr double kOmegaEps = 1e-9;
}  // namespace

DiffDriveModel::DiffDriveModel(double track_width, double wheel_radius,
                               double pole_pairs, double internal_gear,
                               double external_gear)
    : track_width_(track_width),
      wheel_radius_(wheel_radius),
      pole_pairs_(pole_pairs),
      total_gear_(internal_gear * external_gear),
      circumference_(kTwoPi * wheel_radius) {
    if (track_width <= 0.0) throw std::invalid_argument("track_width must be positive");
    if (wheel_radius <= 0.0) throw std::invalid_argument("wheel_radius must be positive");
    if (total_gear_ <= 0.0) throw std::invalid_argument("gear ratio must be positive");
}

WheelSpeeds DiffDriveModel::wheel_speeds_from_twist(double v, double omega) const {
    const double half = omega * track_width_ / 2.0;
    return WheelSpeeds{v - half, v + half};
}

Twist DiffDriveModel::twist_from_wheel_speeds(double v_left, double v_right) const {
    const double v = (v_left + v_right) / 2.0;
    const double omega = (v_right - v_left) / track_width_;
    return Twist{v, omega};
}

double DiffDriveModel::erpm_from_wheel_speed(double v_wheel) const {
    const double wheel_rpm = v_wheel / circumference_ * 60.0;
    return wheel_rpm * total_gear_ * pole_pairs_;
}

double DiffDriveModel::wheel_speed_from_erpm(double erpm) const {
    const double wheel_rpm = erpm / pole_pairs_ / total_gear_;
    return wheel_rpm / 60.0 * circumference_;
}

Pose DiffDriveModel::step_pose(const Pose& pose, double v, double omega,
                               double dt) const {
    double x = pose.x;
    double y = pose.y;
    double theta = pose.heading;
    if (std::abs(omega) < kOmegaEps) {
        x += v * std::cos(theta) * dt;
        y += v * std::sin(theta) * dt;
    } else {
        const double radius = v / omega;
        const double dtheta = omega * dt;
        x += radius * (std::sin(theta + dtheta) - std::sin(theta));
        y -= radius * (std::cos(theta + dtheta) - std::cos(theta));
        theta += dtheta;
    }
    return Pose{x, y, wrap_angle(theta)};
}

}  // namespace backtrack

File: backtrack_demo/backtrack/kinematics.py
"""Differential-drive (skid-steer) kinematics and VESC ERPM conversions.

Models the target platform from the TZ: a skid-steer UGV whose wheels are
driven by VESC-based ESCs (FSESC 75450). Wheel speed is observed on the CAN
bus as motor ERPM; this module converts between the unicycle twist (v, omega),
the per-wheel linear speeds, and ERPM.
"""

import math

from .geometry import wrap_angle


class DiffDriveModel:
    """Kinematic model of a skid-steer rover.

    track_width : distance between left/right wheel contact lines [m]
    wheel_radius: drive wheel radius [m]
    pole_pairs  : motor pole pairs (ERPM = mechanical_rpm * pole_pairs)
    gear_ratio  : motor revolutions per wheel revolution
    """

    def __init__(self, track_width=0.5, wheel_radius=0.165,
                 pole_pairs=15, gear_ratio=1.0):
        if track_width <= 0:
            raise ValueError("track_width must be positive")
        if wheel_radius <= 0:
            raise ValueError("wheel_radius must be positive")
        self.track_width = track_width
        self.wheel_radius = wheel_radius
        self.pole_pairs = pole_pairs
        self.gear_ratio = gear_ratio
        self._circumference = 2.0 * math.pi * wheel_radius

    # -- twist <-> wheel speeds -------------------------------------------
    def wheel_speeds_from_twist(self, v, omega):
        """(linear v, yaw omega) -> (left, right) wheel linear speed [m/s]."""
        half = omega * self.track_width / 2.0
        return v - half, v + half

    def twist_from_wheel_speeds(self, v_left, v_right):
        """(left, right) wheel linear speed -> (linear v, yaw omega)."""
        v = (v_left + v_right) / 2.0
        omega = (v_right - v_left) / self.track_width
        return v, omega

    # -- wheel speed <-> VESC ERPM ----------------------------------------
    def erpm_from_wheel_speed(self, v_wheel):
        """Wheel ground speed [m/s] -> motor electrical RPM."""
        wheel_rpm = v_wheel / self._circumference * 60.0
        return wheel_rpm * self.gear_ratio * self.pole_pairs

    def wheel_speed_from_erpm(self, erpm):
        """Motor electrical RPM -> wheel ground speed [m/s]."""
        wheel_rpm = erpm / self.pole_pairs / self.gear_ratio
        return wheel_rpm / 60.0 * self._circumference

    # -- pose integration --------------------------------------------------
    def step_pose(self, pose, v, omega, dt):
        """Exact arc integration of a unicycle pose.

        pose: (x, y, heading) -> returns new (x, y, heading).
        """
        x, y, theta = pose
        if abs(omega) < 1e-9:
            x += v * math.cos(theta) * dt
            y += v * math.sin(theta) * dt
        else:
            radius = v / omega
            dtheta = omega * dt
            x += radius * (math.sin(theta + dtheta) - math.sin(theta))
            y -= radius * (math.cos(theta + dtheta) - math.cos(theta))
            theta = theta + dtheta
        return x, y, wrap_angle(theta)
